### PR TITLE
ARROW-13220: [C++] Implement 'choose' function

### PR DESCRIPTION
**Pull request by lidavidm**
Also makes two changes:
- MakeArrayFromScalar now works with Decimal256 and FixedSizeBinaryType
- VarArgs(min_args=N) for N>1 now works (before different assertions/runtime checks would trigger). Though, might it make sense to update KernelSignature to reflect min_args?

### CODE REVIEW

**Comment by lidavidm** [comment]
@ursabot please benchmark lang=C++

**Comment by lidavidm** [comment]
This needs a lot of merge conflict resolving, so moving to draft for now until #10608 is also merged.

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 1]
```diff
@@ -1606,6 +1620,203 @@ struct CoalesceFunctor<Type, enable_if_base_binary<Type>> {
...
+  ArrayData* output = out->mutable_array();
+  const int64_t out_offset = output->offset;
+  // Need a null bitmap if any input has nulls
+  uint8_t* out_valid = nullptr;
+  if (std::any_of(batch.values.begin(), batch.values.end(),
+                  [](const Datum& d) { return d.null_count() > 0; })) {
+    out_valid = output->buffers[0]->mutable_data();
+  } else {
+    BitUtil::SetBitsTo(output->buffers[0]->mutable_data(), out_offset, batch.length,
+                       true);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:1665]: Also set `output->null_count` to 0?

File: cpp/src/arrow/compute/kernels/scalar_if_else_test.cc [hunk 2]
```diff
@@ -1040,5 +1040,192 @@ TEST(TestCoalesce, FixedSizeBinary) {
...
+  CheckScalar("choose", {ScalarFromJSON(int64(), "1"), values1, values2}, values2);
+  CheckScalar("choose", {ScalarFromJSON(int64(), "null"), values1, values2}, nulls);
+  auto scalar1 = ScalarFromJSON(type, "42");
+  CheckScalar("choose", {ScalarFromJSON(int64(), "0"), scalar1, values2},
+              *MakeArrayFromScalar(*scalar1, 5));
+  CheckScalar("choose", {ScalarFromJSON(int64(), "1"), scalar1, values2}, values2);
+  CheckScalar("choose", {ScalarFromJSON(int64(), "null"), values1, values2}, nulls);
+  auto scalar_null = ScalarFromJSON(type, "null");
+  CheckScalar("choose", {ScalarFromJSON(int64(), "0"), scalar_null, values2},
+              *MakeArrayOfNull(type, 5));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:1070]: Can you also test where the indices are an array but one of the values is a scalar?

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 3]
```diff
@@ -1606,6 +1620,203 @@ struct CoalesceFunctor<Type, enable_if_base_binary<Type>> {
...
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch.values[0].is_scalar()) {
+      const auto& index_scalar = *batch[0].scalar();
+      if (!index_scalar.is_valid) {
+        if (out->is_array()) {
+          auto null_scalar = MakeNullScalar(out->type());
+          ARROW_ASSIGN_OR_RAISE(
+              auto temp_array,
+              MakeArrayFromScalar(*null_scalar, batch.length, ctx->memory_pool()));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:1720]: Why not reuse `MakeArrayOfNull`?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 4]
```diff
@@ -1723,6 +1963,22 @@ void RegisterScalarIfElse(FunctionRegistry* registry) {
...
+    auto func = std::make_shared<ChooseFunction>("choose", Arity::VarArgs(/*min_args=*/2),
+                                                 &choose_doc);
+    AddPrimitiveChooseKernels(func, NumericTypes());
+    AddPrimitiveChooseKernels(func, TemporalTypes());
+    AddPrimitiveChooseKernels(func,
+                              {boolean(), null(), day_time_interval(), month_interval()});
+    AddChooseKernel(func, Type::FIXED_SIZE_BINARY,
+                    ChooseFunctor<FixedSizeBinaryType>::Exec);
+    AddChooseKernel(func, Type::DECIMAL128, ChooseFunctor<Decimal128Type>::Exec);
+    AddChooseKernel(func, Type::DECIMAL256, ChooseFunctor<Decimal256Type>::Exec);
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:1976]: This unfortunately gets a little annoying as the scalar representations are different (Buffer vs Decimal{128,256}).

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:1988]: Ah, too bad :-(

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:1673]: Even if we start with all-valid, we might select a null later on, so we don't truly know the final null count at this point.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:1673]: I see, thanks.

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**